openedx/core/lib/studio_entities.py

```diff
@@ -0,0 +1,64 @@
+"""Studio entity plugin manager and API."""
+import abc
+
+from openedx.core.lib.api.plugins import PluginManager
+
+
+class StudioEntityPluginManager(PluginManager):
+    """Manager for all available Studio entities.
+
+    Examples of Studio entities include Courses, Libraries, and Programs. All Studio
+    entities should implement `StudioEntity`.
+    """
+    NAMESPACE = 'openedx.studio_entity'
+
+    @classmethod
+    def get_enabled_entities(cls):
+        """Returns a list of enabled Studio entities."""
+        entities = cls.get_available_plugins()
+        enabled_entities = [entity for entity in entities.viewvalues() if entity.is_enabled()]
+
+        return enabled_entities
+
+
+class StudioEntity(object):
+    """Abstract class used to represent Studio entities.
+
+    Examples of Studio entities include Courses, Libraries, and Programs.
+    """
+    __metaclass__ = abc.ABCMeta
```

Comment:
Oh, that's true. I had forgotten that wrinkle, where platform is involved.
Although, as long as the code using the entry-point doesn't check the
subclassing, you don't actually have to import platform, just adhere to the
duck-type.
On Oct 20, 2015 3:25 PM, "Andy Armstrong" <notifications@github.com> wrote:

> In openedx/core/lib/studio_entities.py
> <https://github.com/edx/edx-platform/pull/10253#discussion_r42541406>:
>
> > +
> > +    @classmethod
> > +    def get_enabled_entities(cls):
> > +        """Returns a list of enabled Studio entities."""
> > +        entities = cls.get_available_plugins()
> > +        enabled_entities = [entity for entity in entities.viewvalues() if entity.is_enabled()]
> > +
> > +        return enabled_entities
> > +
> > +
> > +class StudioEntity(object):
> > +    """Abstract class used to represent Studio entities.
> > +
> > +    Examples of Studio entities include Courses, Libraries, and Programs.
> > +    """
> > +    __metaclass__ = abc.ABCMeta
>
> I'm confused :-) Isn't it the case that you are both using Stevedore and
> requiring that each entry point subclass your abstract class. The challenge
> is that your repo will have to import the abstract class from edx-platform
> in order to extend it when it wants to provide a new extension.
> @cpennington <https://github.com/cpennington>, isn't that what you wanted
> to avoid? It will make it hard to write tests as the abstract class won't
> be available in your repo.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/10253/files#r42541406>.
>